storagenode/piecestore/endpoint.go

```diff
@@ -607,48 +598,54 @@ func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainReques
 		return nil, status.Error(codes.InvalidArgument, Error.Wrap(err).Error())
 	}
 
-	const limit = 1000
-	offset := 0
 	numDeleted := 0
-	hasMorePieces := true
 
-	for hasMorePieces {
-		// subtract some time to leave room for clock difference between the satellite and storage node
-		createdBefore := retainReq.GetCreationDate().Add(-endpoint.config.RetainTimeBuffer)
+	// subtract some time to leave room for clock difference between the satellite and storage node
+	createdBefore := retainReq.GetCreationDate().Add(-endpoint.config.RetainTimeBuffer)
+
+	endpoint.log.Info("Prepared to run a Retain request.",
+		zap.Time("createdBefore", createdBefore),
+		zap.Int64("filterSize", filter.Size()),
+		zap.String("satellite", peer.ID.String()))
 
-		pieceIDs, err := endpoint.pieceinfo.GetPieceIDs(ctx, peer.ID, createdBefore, limit, offset)
+	err = endpoint.store.ForAllPieceIDsOwnedBySatellite(ctx, peer.ID, func(access pieces.StoredPieceAccess) error {
+		// We call Gosched() when done because the GC process is expected to be long and we want to keep it at low priority,
+		// so other goroutines can continue serving requests.
+		defer runtime.Gosched()
+		mTime, err := access.ModTime(ctx)
 		if err != nil {
-			return nil, status.Error(codes.Internal, Error.Wrap(err).Error())
+			endpoint.log.Error("failed to determine mtime of blob", zap.Error(err))
+			// but continue iterating.
+			return nil
 		}
-		for _, pieceID := range pieceIDs {
-			if !filter.Contains(pieceID) {
-				endpoint.log.Sugar().Debugf("About to delete piece id (%s) from satellite (%s). RetainStatus: %s", pieceID.String(), peer.ID.String(), endpoint.config.RetainStatus.String())
-
-				// if retain status is enabled, delete pieceid
-				if endpoint.config.RetainStatus == RetainEnabled {
-					if err = endpoint.store.Delete(ctx, peer.ID, pieceID); err != nil {
-						endpoint.log.Error("failed to delete a piece", zap.Error(err))
-						// continue because if we fail to delete from file system,
-						// we need to keep the pieceinfo so we can delete next time
-						continue
-					}
-					if err = endpoint.pieceinfo.Delete(ctx, peer.ID, pieceID); err != nil {
-						endpoint.log.Error("failed to delete piece info", zap.Error(err))
-					}
+		if !mTime.Before(createdBefore) {
+			return nil
+		}
```

Comment:
I'm wondering about the `mTime` accuracy in this check and whether it matters. If the mod time was reset (so it's much later than the actual creation time of the blob) then I suppose this means we could retain some files that should in fact be garbage collected? If that were to happen, would they eventually be deleted in a future garbage collection call?

Reply:
I feel like I remember some weirdness with modified / creation time on files. e.g. https://apenwarr.ca/log/20181113
Similarly, we should consider that at somepoint someone will copy blobs from one hard-drive another, for some reason... which could change modtime/creationtime on the file itself.

In general the GC should apply some threshold of ~1h or more for time drift issues.

Reply:
If that's the case, then it seems likely it would be caught in a future gc call and therefore this probably won't hurt much?

Reply:
Yes, when assuming the mTime is roughly correct.

Reply:
Yes, that is correct. I had a huge comment discussing the exact math of when this would be a problem, given assumptions about system clock drift on participating nodes (answer: when `((fileStoreTime < (GCFilterStartTime + clockDrift_satellite - clockDrift_node - RetainTimeBuffer)) AND (pieceCreateTime > (GCFilterStartTime + clockDrift_satellite - clockDrift_uplink - RetainTimeBuffer)))`) but I think it really was more confusing than helpful. As long as SNOs don't force blob file mtimes to be before their actual mtimes, for some crazy reason, the _worst_ that could happen here would be leaving a piece to be deleted in the very next `Retain()` call. And even if they _did_ set their blob mtimes back to the past, the affected pieces would have to be _very_ recent (so as not to be included in the GC filter) and the new mtime would have to be within the last hour in order to make a piece get deleted incorrectly.

The performance gain from using stat() instead of opening every piece is just too huge not to use it, especially since "don't randomly change the mtimes on the data files owned by the storagenode to weird random times in the past" is a pretty reasonable operational rule.

Reply:
I agree, that makes sense. I hadn't considered the case where they set the mTime to the past (I didn't know that was possible), was only worried about the case where it was much more recent than the actual creation time. But since that would simply mean files hang around a little longer before they are actually garbage-collected in the future, that sounds great.

Reply:
@egonelbre missed your comment before. Yes, mtime is very difficult to work with and is prone to unexpected changes. However, it is reasonable to expect that if mtime does change unexpectedly, it will always increase and never decrease.

The GC system does already have a threshold of `RetainTimeBuffer` to allow for clock drift issues, which is 1h by default, but I would prefer to have something more like `3h`, simply because there is not much downside to being a little more conservative. Should I change the default for that option?

Reply:
I wonder whether we should have a canary file to detect the issues of time moving backwards significantly?

Sure, but in a separate PR.

Reply:
I don't think that would be worth the trouble and the special-casing code, because (a) It really is a reasonable requirement in my experience, even dealing with end users, that the mtimes of internal storage files not artificially be changed to be in the past, and (b) most situations a canary checking system could detect would not really be problematic, and the very narrow sliver of cases that are problematic would be more difficult to catch in time before the damage was done, unless we were to stat the canary very frequently.